### PR TITLE
Partitionable with common attributes

### PR DESCRIPTION
**Pull request by johnbelamaric**
This is an evolution of the partitionable model defined in #27, which moves common attributes up to the pool level to reduce the size of the objects.

### CODE REVIEW

File: dra-evolution/pkg/api/capacity_types.go [hunk 1]
```diff
@@ -68,6 +68,11 @@ type ResourcePoolSpec struct {
 	// +optional
 	SharedCapacity []SharedCapacity `json:"sharedCapacity,omitempty"`
 
+	// Attributes contains common device attributes that are the same
+	// for all devices in the pool, unless a device specifically over
```
Comment (pohly) [inline dra-evolution/pkg/api/capacity_types.go:72]: "the pool" or "the slice"?

The prototype is not up-to-date here :cry: We should update it to match the current KEP before doing further prototyping.

**Comment by pohly** [comment]
> to reduce the size of the objects

That helps reduce the size on average, but for the worst-case analysis which determines the limits of the slices we have to assume that the new `Attributes` is fully-populated *and* all devices have the maximum number of attributes.

File: dra-evolution/pkg/api/capacity_types.go (hunk 1 above)
Comment (johnbelamaric) [reply dra-evolution/pkg/api/capacity_types.go:72]: Yeah. I didn't want to do that yet, I figured it could be done and I could rebase...

**Comment by johnbelamaric** [comment]
> > to reduce the size of the objects
> 
> That helps reduce the size on average, but for the worst-case analysis which determines the limits of the slices we have to assume that the new `Attributes` is fully-populated _and_ all devices have the maximum number of attributes.

Yes. True.

**Comment by johnbelamaric** [comment]
I am making several options, see comment in #20

File: dra-evolution/testdata/pools-two-nodes-dgxa100.yaml [hunk 2]
```diff
@@ -0,0 +1,7695 @@
...
+    - name: product-name
+      string: Mock NVIDIA A100-SXM4-40GB
+    devices:
+    - attributes:
+      - name: uuid
+        string: GPU-cdd0dd02-b5dd-4ae9-9c5a-c76b19086eb4
+      - bool: true
+        name: mig-capable
+      - bool: false
+        name: mig-enabled
```
Comment (thockin) [inline dra-evolution/testdata/pools-two-nodes-dgxa100.yaml]: NOTE: we should establish core guidelines - are attributes expected to be mutable or immutable?  Do we encourage drivers to embed status here or are attributes only *intrinsic* things about the device?  This one feels like status.

File: dra-evolution/testdata/pools-two-nodes-dgxa100.yaml [hunk 3]
```diff
@@ -0,0 +1,7695 @@
...
+      - capacity: 4864Mi
+        name: gpu-0-memory
+      - capacity: "1"
+        name: gpu-0-memory-slices-0
+    - attributes:
+      - name: mig-profile
+        string: 1g.5gb+me
+      name: gpu-0-mig-1g.5gb-me-1
+      sharedCapacityConsumed:
+      - capacity: "14"
```
Comment (thockin) [inline dra-evolution/testdata/pools-two-nodes-dgxa100.yaml:294]: "capacity" is a producer term, not a consumer term.  This should be "quantity" or "amount" or "request" or something.

File: dra-evolution/testdata/pools-two-nodes-dgxa100.yaml [hunk 4]
```diff
@@ -0,0 +1,7695 @@
...
+        name: gpu-0-copy-engines
+      - capacity: "1"
+        name: gpu-0-decoders
+      - capacity: "1"
+        name: gpu-0-jpeg-engines
+      - capacity: "1"
+        name: gpu-0-ofa-engines
+      - capacity: 4864Mi
+        name: gpu-0-memory
+      - capacity: "1"
```
Comment (thockin) [inline dra-evolution/testdata/pools-two-nodes-dgxa100.yaml]: MAYBE (really not sure) we should say that if `capacity` is omitted here, it means "all"?  Then it is sort of a fancy mutex for "partitions" and an quantity for fungibles.

Maybe not worthwhile

**Review by thockin — COMMENTED** [review]
I think it is clear that shared attributes is an appropriate optimization to include.

I think this further emphasizes that some form of nesting would (IMO) make other optimizations possible.  There are attributes that apply to all "cards" (e.g. vendor).  There are some attributes that apply to all GPUs (MIG or not) on a single card (e.g. the card UUID, for matching).  There are some attributes that apply to each leaf device.  Why are we shoehorning that into 2 levels?

**Comment by johnbelamaric** [comment]
Option 4 has some nesting. That's #31. It is much more efficient than this one.

**Comment by johnbelamaric** [comment]
We could do more levels. Not clear the payoff is there.

**Comment by johnbelamaric** [comment]
Option 3 makes common attributes. Option 4 makes common attributes AND common partition "map".

File: dra-evolution/testdata/pools-two-nodes-dgxa100.yaml (hunk 2 above)
Comment (klueska) [reply dra-evolution/testdata/pools-two-nodes-dgxa100.yaml]: This should not be an attribute. MIG capable yes, enabled no. This is leftover from the structure I used to represent all of this in classic DRA. It is very compact, but also not generalizable.

File: dra-evolution/testdata/pools-two-nodes-dgxa100.yaml (hunk 3 above)
Comment (klueska) [reply dra-evolution/testdata/pools-two-nodes-dgxa100.yaml:294]: It reuses the same struct as that used to declare the capacity. We would need a new struct then (which seems fine).

**Comment by klueska** [comment]
I 100% agree we should have shared attributes. My first prototype had it, but at the time you said „let’s not prematurely optimize for size“ so we dropped it.

I’m still not sold on nesting though. I also had it originally (in my „recursive“ device model), but you all (rightly) convinced me to drop it. And now after working with flat devices and updating both the example driver and the NVIDIA GPU driver to  adhere to them, I’m really happy with the flexibility that a flat model brings us.

I really don’t think it buys us much to have nesting and I have a strong feeling will come back to bite us fairly quickly.

**Comment by thockin** [comment]
I was continuing on the KEP PR, but detoured to these option, so I will copy a comment:

I just find it super weird to have gpu1's shared items consumable from gpu0.  This is the thing that is setting me on edge.  That implies to me a level of fungibility which doesn't exist.  There *is* a grouping that is smaller than a ResourceSlice but bigger than a Device, and we are not modelling it.  Call it a "card" for the moment.  A slice doesn't have shared resources, a "card" does.  Now you'll probably tell me "actually, one card can borrow resources on another card".  In fact, I can already see the (hypothetical) use-case for a channelized `<something>` which can be effectively RAID'ed into a larger logical device.  But that's not this, and (TTBOMK) that doesn't exist yet.

I really can see both sides, and I don't mean to be dogmatic.  It just smells funny.  Let's keep the conversation overall moving forward, and if this is all that's left, we can hash it out.

>  My first prototype had it, but at the time you said „let’s not prematurely optimize for size“ so we dropped it.

Yeah, we dropped a LOT to get the baseline, and bringing partitions back makes it clear to me that this is one piece that really does make sense.

File: dra-evolution/testdata/pools-two-nodes-dgxa100.yaml [hunk 5]
```diff
@@ -0,0 +1,7695 @@
...
+      - bool: true
+        name: mig-capable
+      - bool: false
+        name: mig-enabled
+      name: gpu-0
+      sharedCapacityConsumed:
+      - capacity: 40Gi
+        name: gpu-0-memory
+    - attributes:
+      - name: mig-profile
```
Comment (thockin) [inline dra-evolution/testdata/pools-two-nodes-dgxa100.yaml:33]: We should include in this YAML the requirement that all MIG partitions on a "card" have a "parent UUID" attribute with the same value (for matching).  I think that would highlight the "hidden" nesting level, even if we ultimately choose not to model it.

**Pull request closed without merging**